database/src/config.rs

```diff
@@ -16,151 +26,355 @@ use crate::{constants::DATABASE_FILENAME, resize::ResizeAlgorithm};
         let (db_directory, db_file) = Self::return_db_dir_and_file(db_directory);
         Self {
             db_directory,
             db_file,
-            sync_mode: SyncMode::Fastest,
+            sync_mode: SyncMode::Fast,
             reader_threads: ReaderThreads::OnePerThread,
             resize_algorithm: ResizeAlgorithm::new(),
         }
     }
 
-    /// TODO
-    pub fn low_power<P: AsRef<Path>>(db_directory: Option<P>) -> Self {
+    /// Create a [`Config`] with the lowest performing,
+    /// but also least resource-intensive settings.
+    ///
+    /// Good default for resource-limited machines, e.g. a cheap VPS.
+    ///
+    /// # `db_directory`
+    /// If this is `Some`, it will be used as the
+    /// directory that contains all database files.
+    ///
+    /// If `None`, it will use the default directory [`cuprate_database_dir`].
+    pub fn low_power(db_directory: Option<PathBuf>) -> Self {
         let (db_directory, db_file) = Self::return_db_dir_and_file(db_directory);
         Self {
             db_directory,
             db_file,
-            sync_mode: SyncMode::Safe,
+            sync_mode: SyncMode::FastThenSafe,
             reader_threads: ReaderThreads::One,
             resize_algorithm: ResizeAlgorithm::new(),
         }
     }
 
+    /// Return the absolute [`Path`] to the database directory.
+    ///
+    /// This will be the `db_directory` given
+    /// (or default) during [`Config`] construction.
+    pub const fn db_directory(&self) -> &Cow<'_, Path> {
+        &self.db_directory
+    }
+
     /// Return the absolute [`Path`] to the database data file.
-    pub fn db_file_path(&self) -> &Path {
+    ///
+    /// This will be based off the `db_directory` given
+    /// (or default) during [`Config`] construction.
+    pub const fn db_file(&self) -> &Cow<'_, Path> {
         &self.db_file
     }
 }
 
 impl Default for Config {
+    /// Same as `Self::new(None)`.
+    ///
+    /// ```rust
+    /// # use cuprate_database::config::*;
+    /// assert_eq!(Config::default(), Config::new(None));
+    /// ```
     fn default() -> Self {
-        Self::new(None::<&'static Path>)
+        Self::new(None)
     }
 }
 
 //---------------------------------------------------------------------------------------------------- SyncMode
-/// TODO
-#[derive(Copy, Clone, Default, PartialEq, PartialOrd, Eq, Ord, Hash)]
+/// Disk synchronization mode.
+///
+/// This controls how/when the database syncs its data to disk.
+///
+/// Regardless of the variant chosen, dropping [`Env`](crate::Env)
+/// will always cause it to fully sync to disk.
+///
+/// # Sync vs Async
+/// All invariants except [`SyncMode::Fast`] are `synchronous`,
+/// as in the database will wait until the sync is finished before continuing.
+///
+/// `SyncMode::Fast` is `asynchronous`, meaning it will _NOT_
+/// wait until the sync is done before continuing. It will immediately
+/// move onto the next operation even if the database/OS has not responded.
```

Comment:
This would not be safe, when a write happens a new reader must be able to get the new data. This means we would need the DB to respond to the writer so we know that when we call a write, when that call finishes we can open a new read tx and get this data.

All these config options should be doing is changing how we open the Env

Reply:
Sorry, these docs were a little misleading...

This variant will just open LMDB with [`MDB_NOSYNC`](http://www.lmdb.tech/doc/group__mdb__env.html#ga5791dd1adb09123f82dd1f331209e12e) and never call `sync()`. This matches `fastest:async` mode in `monerod`. The readers will still be able to see writes regardless of this, it's referring to _disk_ syncs, not the database data itself so:

```rust
db.put("key", value);
db.get("key"); // <- will get value even with `Fast`, probably from memory directly
```